airbyte-server/src/main/java/io/airbyte/server/ConnectorDefinitionSpecBackfiller.java

```diff
@@ -0,0 +1,263 @@
+import io.airbyte.config.SourceConnection;
+import io.airbyte.config.StandardDestinationDefinition;
+import io.airbyte.config.StandardSourceDefinition;
+import io.airbyte.config.StandardSync;
+import io.airbyte.config.persistence.ConfigNotFoundException;
+import io.airbyte.config.persistence.ConfigPersistence;
+import io.airbyte.config.persistence.ConfigRepository;
+import io.airbyte.protocol.models.ConnectorSpecification;
+import io.airbyte.scheduler.client.SynchronousResponse;
+import io.airbyte.scheduler.client.SynchronousSchedulerClient;
+import io.airbyte.server.converters.JobConverter;
+import io.airbyte.validation.json.JsonValidationException;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// todo (lmossman) - delete this class after the faux major version bump, along with making the spec
+// field required on the definition structs
+public class ConnectorDefinitionSpecBackfiller {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ConnectorDefinitionSpecBackfiller.class);
+
+  /**
+   * Check that each spec in the database has a spec. If it doesn't, add it. If it can't be added,
+   * delete the connector definition or fail according to the VERSION_0_31_0_FORCE_UPGRADE env var.
+   * The goal is to try to end up in a state where all definitions in the db contain specs, or fail
+   * otherwise.
+   *
+   * @param configRepository - access to the db
+   * @param database - access to the db at a lower level (for deleting connections and syncs)
+   * @param schedulerClient - scheduler client so that specs can be fetched as needed
+   * @param trackingClient - tracking client for reporting failures to Segment
+   * @param configs - for retrieving various configs (env vars, worker environment, logs)
+   */
+  @VisibleForTesting
+  static void migrateAllDefinitionsToContainSpec(final ConfigRepository configRepository,
+                                                 final ConfigPersistence database,
+                                                 final ConfigPersistence seed,
+                                                 final SynchronousSchedulerClient schedulerClient,
+                                                 final TrackingClient trackingClient,
+                                                 final Configs configs)
+      throws IOException, JsonValidationException, ConfigNotFoundException {
+    final List<String> failedBackfillImages = Lists.newArrayList();
+
+    final JobConverter jobConverter = new JobConverter(configs.getWorkerEnvironment(), configs.getLogConfigs());
+
+    final Set<String> connectorReposInUse = database.getConnectorReposInUse();
+
+    final List<String> seedSourceRepos = seed.listConfigs(ConfigSchema.STANDARD_SOURCE_DEFINITION, StandardSourceDefinition.class)
```

Comment:
should this be a set instead of a list? i think you just use it to check membership.

Reply:
Yep set should work fine here, done